staging/src/k8s.io/apiserver/pkg/server/options/etcd.go

```diff
@@ -204,6 +205,20 @@ func (s *EtcdOptions) addEtcdHealthEndpoint(c *server.Config) error {
 	c.HealthzChecks = append(c.HealthzChecks, healthz.NamedCheck("etcd", func(r *http.Request) error {
 		return healthCheck()
 	}))
+
+	if s.EncryptionProviderConfigFilepath != "" {
+		healthzConfigs, err := encryptionconfig.GetKMSPluginProbes(s.EncryptionProviderConfigFilepath)
+		if err != nil {
+			return err
+		}
+
+		for i, h := range healthzConfigs {
+			c.HealthzChecks = append(c.HealthzChecks, healthz.NamedCheck(fmt.Sprintf("kms-plugin-%d", i), func(r *http.Request) error {
+				return h.Check()
```

Comment:
the `h` captured here by the closure is incorrect, it means if you have multiple checks, they will all test the last iterated config

Reply:
add a test with at least two configs and verify the checks exercise the correct KMS

Reply:
isn't the first KMS in the list used to encrypt, and subsequent ones to decrypt? would we ever encounter a case where a later KMS would only permit decrypting, so the encrypt("ping") check would fail?

Reply:
Fixed the closure bug.
Added integration test for multiple providers' scenario (see, kms_treansformer_test.go TestHealthz.

Yes, theoretically, the scenario you described in the third comment is possible. KMS Administrator may remove "encrypt" IAM privilege from the service account under which the kms-plugin runs once the corresponding provider is moved into the second place - used only for decryption.
In practice, this is unlikely, but we should probably document this.
Also, there is no generic way to test decryption without providing a valid encryption payload.
I think the long term solution here is to move towards gRPC Health Checking Protocol
https://github.com/grpc/grpc/blob/master/doc/health-checking.md
KMS-Plugin developers may have vendor specific way to ascertain the status of the plugin based on the requested operation type.